File: apps/desktop/src/features/chart/optionsAnalytics/optionsAnalyticsPresentation.ts
import type { OptionsAnalyticsSnapshot } from '@0dtetrader/shared-types';
import type { OptionsAnalyticsSettings } from './optionsAnalyticsSettings';
import { selectVisibleOptionsAnalyticsStrikes } from './optionsAnalyticsGeometry';

export interface OptionsAnalyticsLiquidityPresentation {
  callBidSize: number | null;
  callAskSize: number | null;
  putBidSize: number | null;
  putAskSize: number | null;
  callOpenInterest: number | null;
  putOpenInterest: number | null;
  callVolume: number | null;
  putVolume: number | null;
  callRelativeSpread: number | null;
  putRelativeSpread: number | null;
  callRoundTripCost: number | null;
  putRoundTripCost: number | null;
}

export interface OptionsAnalyticsStrikePresentation {
  strike: number;
  callImpliedVolatility: number | null;
  putImpliedVolatility: number | null;
  callDelta: number | null;
  putDelta: number | null;
  callDeltaNotional: number | null;
  putDeltaNotional: number | null;
  callGammaExposure: number;
  putGammaExposure: number;
  callScale: number;
  putScale: number;
  callDirection: 'right';
  putDirection: 'left';
  profileImportance: number;
  grossGammaExposure: number;
  totalOpenInterest: number;
  markedOiValue: number | null;
  callMarkedOiValue: number | null;
  putMarkedOiValue: number | null;
  callMarkedOiScale: number;
  putMarkedOiScale: number;
  liquidity: OptionsAnalyticsLiquidityPresentation | null;
  dealerProxyGammaExposure: number | null;
}

export interface OptionsAnalyticsPresentation {
  strikes: OptionsAnalyticsStrikePresentation[];
  allStrikes: OptionsAnalyticsStrikePresentation[];
  showGammaProfile: boolean;
  showMarkedOi: boolean;
  showLiquidity: boolean;
  profileStrikeCount: number;
  impliedRange: OptionsAnalyticsSnapshot['impliedRange'];
  callWall: number | null;
  putWall: number | null;
  dealerProxy: NonNullable<OptionsAnalyticsSnapshot['scenarios']['callPutDealerProxy']> | null;
  structureLine: string;
  qualityLines: string[];
  visibleQualityLines: string[];
  accessibleSummary: string;
}

function formatPrice(value: number): string {
  return value.toFixed(2);
}

export function formatCompactDollars(value: number): string {
  const absolute = Math.abs(value);
  const sign = value < 0 ? '-' : '';
  if (absolute >= 1e9) return `${sign}$${(absolute / 1e9).toFixed(1)}B`;
  if (absolute >= 1e6) return `${sign}$${(absolute / 1e6).toFixed(1)}M`;
  if (absolute >= 1e3) return `${sign}$${(absolute / 1e3).toFixed(0)}K`;
  return `${sign}$${absolute.toFixed(0)}`;
}

function formatOptionalCompactDollars(value: number | null): string {
  return value === null ? 'unavailable' : formatCompactDollars(value);
}

function formatTimestamp(iso: string | null): string {
  if (iso === null) return 'unavailable';
  const parsed = Date.parse(iso);
  return Number.isFinite(parsed)
    ? `${new Date(parsed).toISOString().slice(11, 19)} UTC`
    : 'unavailable';
}

function formatCurrentAge(nowMs: number, sourceAt: string | null): string {
  if (sourceAt === null) return 'unavailable';
  const source = Date.parse(sourceAt);
  if (!Number.isFinite(source)) return 'unavailable';
  return `${Math.max(0, Math.round((nowMs - source) / 1_000))}s`;
}

function sumNullable(first: number | null | undefined, second: number | null | undefined) {
  if (first === null || first === undefined) {
    return second === null || second === undefined ? null : second;
  }
  return first + (second ?? 0);
}

function withProfileImportance(
  strikes: OptionsAnalyticsStrikePresentation[],
  settings: OptionsAnalyticsSettings,
): OptionsAnalyticsStrikePresentation[] {
  const metrics: Array<(strike: OptionsAnalyticsStrikePresentation) => number> = [];
  if (settings.showGammaProfile) {
    metrics.push((strike) => Math.abs(strike.grossGammaExposure));
  }
  if (settings.showMarkedOi) {
    metrics.push((strike) => strike.markedOiValue ?? 0);
  }
  if (settings.showLiquidity) {
    metrics.push(
      (strike) =>
        (strike.liquidity?.callBidSize ?? 0) +
        (strike.liquidity?.callAskSize ?? 0) +
        (strike.liquidity?.putBidSize ?? 0) +
        (strike.liquidity?.putAskSize ?? 0),
      (strike) =>
        (strike.liquidity?.callOpenInterest ?? 0) + (strike.liquidity?.putOpenInterest ?? 0),
      (strike) => (strike.liquidity?.callVolume ?? 0) + (strike.liquidity?.putVolume ?? 0),
      (strike) =>
        Math.max(
          strike.liquidity?.callRelativeSpread ?? 0,
          strike.liquidity?.putRelativeSpread ?? 0,
        ),
      (strike) =>
        Math.max(strike.liquidity?.callRoundTripCost ?? 0, strike.liquidity?.putRoundTripCost ?? 0),
    );
  }

  const maxima = metrics.map((metric) => Math.max(0, ...strikes.map(metric)));
  return strikes.map((strike) => ({
    ...strike,
    profileImportance: Math.max(
      0,
      ...metrics.map((metric, index) => {
        const maximum = maxima[index] ?? 0;
        return maximum === 0 ? 0 : metric(strike) / maximum;
      }),
    ),
  }));
}

/** Uses the full point-in-time snapshot so viewport changes never resize retained OI bars. */
function withStableMarkedOiScale(
  strikes: OptionsAnalyticsStrikePresentation[],
): OptionsAnalyticsStrikePresentation[] {
  const maximumMarkedOi = Math.max(
    0,
    ...strikes.flatMap((strike) => [strike.callMarkedOiValue ?? 0, strike.putMarkedOiValue ?? 0]),
  );
  const scale = (value: number | null): number =>
    value === null || maximumMarkedOi === 0 ? 0 : Math.sqrt(value / maximumMarkedOi);
  return strikes.map((strike) => ({
    ...strike,
    callMarkedOiScale: scale(strike.callMarkedOiValue),
    putMarkedOiScale: scale(strike.putMarkedOiValue),
  }));
}

export function buildOptionsAnalyticsPresentation(
  snapshot: OptionsAnalyticsSnapshot,
  settings: OptionsAnalyticsSettings,
  nowMs: number = Date.now(),
): OptionsAnalyticsPresentation {
  const proxyByStrike = new Map(
    snapshot.scenarios.callPutDealerProxy?.strikeGammaExposures.map((entry) => [
      entry.strike,
      entry.gammaExposure,
    ]) ?? [],
  );
  const unscoredStrikes = snapshot.strikes.map<OptionsAnalyticsStrikePresentation>((strike) => {
    const callGammaExposure = strike.call?.gammaExposure ?? 0;
    const putGammaExposure = strike.put?.gammaExposure ?? 0;
    return {
      strike: strike.strike,
      callImpliedVolatility: strike.call?.impliedVolatility ?? null,
      putImpliedVolatility: strike.put?.impliedVolatility ?? null,
      callDelta: strike.call?.delta ?? null,
      putDelta: strike.put?.delta ?? null,
      callDeltaNotional: strike.call?.deltaNotional ?? null,
      putDeltaNotional: strike.put?.deltaNotional ?? null,
      callGammaExposure,
      putGammaExposure,
      callScale: 0,
      putScale: 0,
      callDirection: 'right',
      putDirection: 'left',
      profileImportance: 0,
      grossGammaExposure: strike.grossGammaExposure ?? 0,
      totalOpenInterest: strike.totalOpenInterest,
      markedOiValue: settings.showMarkedOi
        ? sumNullable(strike.call?.markedOiValue, strike.put?.markedOiValue)
        : null,
      callMarkedOiValue: settings.showMarkedOi ? (strike.call?.markedOiValue ?? null) : null,
      putMarkedOiValue: settings.showMarkedOi ? (strike.put?.markedOiValue ?? null) : null,
      callMarkedOiScale: 0,
      putMarkedOiScale: 0,
      liquidity: settings.showLiquidity
        ? {
            callBidSize: strike.call?.bidSize ?? null,
            callAskSize: strike.call?.askSize ?? null,
            putBidSize: strike.put?.bidSize ?? null,
            putAskSize: strike.put?.askSize ?? null,
            callOpenInterest: strike.call?.openInterest ?? null,
            putOpenInterest: strike.put?.openInterest ?? null,
            callVolume: strike.call?.volume ?? null,
            putVolume: strike.put?.volume ?? null,
            callRelativeSpread: strike.call?.relativeSpread ?? null,
            putRelativeSpread: strike.put?.relativeSpread ?? null,
            callRoundTripCost: strike.call?.roundTripCost ?? null,
            putRoundTripCost: strike.put?.roundTripCost ?? null,
          }
        : null,
      dealerProxyGammaExposure: settings.showDealerProxy
        ? (proxyByStrike.get(strike.strike) ?? null)
        : null,
    };
  });
  const allStrikes = withStableMarkedOiScale(withProfileImportance(unscoredStrikes, settings));
  const strikes = selectOptionsAnalyticsProfileStrikes(
    allStrikes,
    settings.profileStrikeCount,
    () => true,
  );

  const { scope, quality } = snapshot;
  const coveragePercent = Math.round(quality.coverage.ratio * 100);
  const warningLines =
    quality.warnings.length > 0
      ? quality.warnings.map((warning) => `Warning: ${warning}`)
      : ['Warnings none'];
  const provenanceLines = [
    `Expiration ${scope.expiration} · Root ${scope.rootSymbol} · Settlement ${scope.settlementStyle.toUpperCase()} · ${snapshot.exposureUnit}`,
    `Observed ${formatTimestamp(scope.observedAt)} (age ${formatCurrentAge(nowMs, scope.observedAt)}) · Quote ${formatTimestamp(quality.quoteAsOf)} (age ${formatCurrentAge(nowMs, quality.quoteAsOf)}) · Greeks ${formatTimestamp(quality.greeksAsOf)} (age ${formatCurrentAge(nowMs, quality.greeksAsOf)})`,
    `Feed ${quality.feedMode} · Coverage ${quality.coverage.contractsIncluded}/${quality.coverage.contractsTotal} (${coveragePercent}%) · Status ${quality.status} · Cache ${quality.cacheStatus}`,
    `Version ${quality.calculationVersion} · OI effective ${quality.oiEffectiveDate ?? 'unavailable'}`,
  ];
  const qualityLines = [...provenanceLines, ...warningLines];
  const visibleWarningLines =
    warningLines.length <= 3
      ? warningLines
      : [...warningLines.slice(0, 3), `${warningLines.length - 3} more warnings`];
  const visibleQualityLines = [...provenanceLines, ...visibleWarningLines];
  const structureLine = `Structure gamma C ${formatOptionalCompactDollars(snapshot.structure.callGammaExposure)} · P ${formatOptionalCompactDollars(snapshot.structure.putGammaExposure)} · gross ${formatOptionalCompactDollars(snapshot.structure.grossGammaExposure)} · delta notional C ${formatOptionalCompactDollars(snapshot.structure.callDeltaNotional)} · P ${formatOptionalCompactDollars(snapshot.structure.putDeltaNotional)}`;

  const summaryParts = [...qualityLines, structureLine];
  const impliedRange = settings.showImpliedRange ? snapshot.impliedRange : null;
  if (impliedRange) {
    summaryParts.push(
      `${impliedRange.label} ${formatPrice(impliedRange.lower)} to ${formatPrice(impliedRange.upper)}`,
      `straddle breakevens ${formatPrice(impliedRange.straddleLower)} to ${formatPrice(impliedRange.straddleUpper)}`,
    );
  }
  if (snapshot.structure.callWall !== null) {
    summaryParts.push(`call wall ${formatPrice(snapshot.structure.callWall)}`);
  }
  if (snapshot.structure.putWall !== null) {
    summaryParts.push(`put wall ${formatPrice(snapshot.structure.putWall)}`);
  }
  const dealerProxy = settings.showDealerProxy ? snapshot.scenarios.callPutDealerProxy : null;
  if (dealerProxy) {
    summaryParts.push(`dealer proxy assumption ${dealerProxy.assumption}`);
    summaryParts.push(
      `dealer proxy gamma exposure ${dealerProxy.gammaExposure.toFixed(2)}`,
      `dealer proxy delta notional ${dealerProxy.deltaNotional.toFixed(2)}`,
    );
    summaryParts.push(
      dealerProxy.gammaRoots.length > 0
        ? `dealer proxy roots ${dealerProxy.gammaRoots.map(formatPrice).join(', ')}`
        : 'dealer proxy roots unavailable',
    );
  }

  return {
    strikes,
    allStrikes,
    showGammaProfile: settings.showGammaProfile,
    showMarkedOi: settings.showMarkedOi,
    showLiquidity: settings.showLiquidity,
    profileStrikeCount: settings.profileStrikeCount,
    impliedRange,
    callWall: snapshot.structure.callWall,
    putWall: snapshot.structure.putWall,
    dealerProxy,
    structureLine,
    qualityLines,
    visibleQualityLines,
    accessibleSummary: summaryParts.join('. '),
  };
}

export function scaleOptionsAnalyticsStrikes(
  strikes: OptionsAnalyticsStrikePresentation[],
  normalizationStrikes: OptionsAnalyticsStrikePresentation[] = strikes,
): OptionsAnalyticsStrikePresentation[] {
  const maximumLegGamma = Math.max(
    0,
    ...normalizationStrikes.flatMap((strike) => [
      Math.abs(strike.callGammaExposure),
      Math.abs(strike.putGammaExposure),
    ]),
  );
  const scale = (value: number): number =>
    maximumLegGamma === 0 ? 0 : Math.sqrt(Math.abs(value) / maximumLegGamma);
  return strikes.map((strike) => ({
    ...strike,
    callScale: scale(strike.callGammaExposure),
    putScale: scale(strike.putGammaExposure),
  }));
}

export function selectOptionsAnalyticsProfileStrikes(
  allStrikes: OptionsAnalyticsStrikePresentation[],
  limit: number,
  isVisible: (strike: OptionsAnalyticsStrikePresentation) => boolean,
  score?: (strike: OptionsAnalyticsStrikePresentation) => number,
): OptionsAnalyticsStrikePresentation[] {
  return scaleOptionsAnalyticsStrikes(
    selectVisibleOptionsAnalyticsStrikes(
      allStrikes,
      limit,
      isVisible,
      score ?? ((strike) => strike.profileImportance),
    ),
    allStrikes,
  );
}

File: apps/ios/0dteTrader/Features/Chart/OptionsAnalytics/OptionsAnalyticsPresentation.swift
import Foundation
import UIKit

struct OptionsAnalyticsProfileNormalization: Equatable, Sendable {
    let gammaExposure: Double
    let markedOiValue: Double
}

struct OptionsAnalyticsLegDetail: Equatable, Sendable {
    let label: String
    let value: String
}

enum OptionsAnalyticsPresentation {
    static let callColor = UIColor(red: 34 / 255, green: 224 / 255, blue: 106 / 255, alpha: 1)
    static let putColor = UIColor(red: 1, green: 89 / 255, blue: 105 / 255, alpha: 1)
    static let rangeColor = UIColor(red: 100 / 255, green: 210 / 255, blue: 1, alpha: 1)
    static let markedOiCallColor = UIColor(red: 1, green: 159 / 255, blue: 10 / 255, alpha: 1)
    static let markedOiPutColor = UIColor(red: 1, green: 214 / 255, blue: 10 / 255, alpha: 1)
    static let proxyColor = UIColor(red: 190 / 255, green: 130 / 255, blue: 1, alpha: 1)

    static func notionalText(_ value: Double) -> String {
        guard value != 0 else { return "$0" }
        let magnitude = abs(value)
        let sign = value < 0 ? "-" : "+"
        switch magnitude {
        case 1_000_000_000...:
            return String(format: "%@$%.1fB", sign, rounded(magnitude / 1_000_000_000, digits: 1))
        case 1_000_000...:
            return String(format: "%@$%.1fM", sign, rounded(magnitude / 1_000_000, digits: 1))
        case 1_000...:
            return String(format: "%@$%.0fK", sign, rounded(magnitude / 1_000, digits: 0))
        default:
            return String(format: "%@$%.0f", sign, magnitude)
        }
    }

    private static func optionalNotionalText(_ value: Double?) -> String {
        value.map(notionalText) ?? "unavailable"
    }

    private static func rounded(_ value: Double, digits: Int) -> Double {
        let factor = pow(10, Double(digits))
        return (value * factor).rounded(.toNearestOrAwayFromZero) / factor
    }

    static func magnitudeFraction(value: Double, maximum: Double) -> Double {
        guard maximum > 0, maximum.isFinite, value.isFinite else { return 0 }
        return min(1, sqrt(abs(value) / maximum))
    }

    static func profileNormalization(
        for strikes: [OptionsAnalyticsStrikeDTO]
    ) -> OptionsAnalyticsProfileNormalization {
        let legs = strikes.flatMap { [$0.call, $0.put].compactMap({ $0 }) }
        return OptionsAnalyticsProfileNormalization(
            gammaExposure: legs.lazy.compactMap(\.gammaExposure).map(abs).max() ?? 0,
            markedOiValue: legs.lazy.compactMap(\.markedOiValue).max() ?? 0
        )
    }

    static func selectedProfileStrikes(
        strikes: [OptionsAnalyticsStrikeDTO],
        spot: Double,
        settings: OptionsAnalyticsSettings,
        isVisible: (OptionsAnalyticsStrikeDTO) -> Bool = { _ in true }
    ) -> [OptionsAnalyticsStrikeDTO] {
        let gammaValues = strikes.compactMap(\.grossGammaExposure)
        let markedOiValues = strikes.map(markedOiScore)
        let maximumGamma = gammaValues.max() ?? 0
        let maximumMarkedOi = markedOiValues.max() ?? 0
        let liquidityMetrics: [(OptionsAnalyticsStrikeDTO) -> Double] = [
            quoteSizeScore,
            openInterestScore,
            volumeScore,
            relativeSpreadScore,
            roundTripScore,
        ]
        let liquidityMetricRanges = liquidityMetrics.map { metric in
            let values = strikes.lazy.map(metric)
            return (minimum: values.min() ?? 0, maximum: values.max() ?? 0)
        }

        func normalizedScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
            var scores: [Double] = []
            if settings.showGammaProfile {
                scores.append(maximumGamma > 0 ? (strike.grossGammaExposure ?? 0) / maximumGamma : 0)
            }
            if settings.showMarkedOi {
                scores.append(maximumMarkedOi > 0 ? markedOiScore(strike) / maximumMarkedOi : 0)
            }
            if settings.showLiquidity {
                let liquidityScores = liquidityMetrics.enumerated().map { index, metric in
                    let range = liquidityMetricRanges[index]
                    let width = range.maximum - range.minimum
                    return width > 0 ? (metric(strike) - range.minimum) / width : 0
                }
                scores.append(liquidityScores.max() ?? 0)
            }
            return scores.max() ?? 0
        }

        let candidates = strikes.filter(isVisible)
        return Array(
            candidates.sorted { left, right in
                let leftScore = normalizedScore(left)
                let rightScore = normalizedScore(right)
                if leftScore != rightScore { return leftScore > rightScore }
                let leftDistance = abs(left.strike - spot)
                let rightDistance = abs(right.strike - spot)
                if leftDistance != rightDistance { return leftDistance < rightDistance }
                return left.strike < right.strike
            }.prefix(settings.profileStrikeCount)
        ).sorted { $0.strike < $1.strike }
    }

    static func legDetails(
        side: String,
        leg: OptionsAnalyticsLegDTO
    ) -> [OptionsAnalyticsLegDetail] {
        [
            OptionsAnalyticsLegDetail(label: "\(side) open interest", value: "\(leg.openInterest)"),
            OptionsAnalyticsLegDetail(label: "\(side) volume", value: "\(leg.volume)"),
            OptionsAnalyticsLegDetail(
                label: "\(side) implied volatility",
                value: leg.impliedVolatility.map { String(format: "%.1f%%", $0 * 100) }
                    ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) delta",
                value: leg.delta.map { String(format: "%.4f", $0) } ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) gamma per $1 move",
                value: leg.gamma.map { String(format: "%.6f", $0) } ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) gamma exposure per 1% move",
                value: leg.gammaExposure.map(notionalText) ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) delta notional",
                value: leg.deltaNotional.map(notionalText) ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) quote sizes",
                value: "\(leg.bidSize) bid / \(leg.askSize) ask"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) relative spread",
                value: leg.relativeSpread.map { String(format: "%.1f%%", $0 * 100) } ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) per-contract round trip",
                value: leg.roundTripCost.map { String(format: "$%.2f", $0) } ?? "Unavailable"
            ),
            OptionsAnalyticsLegDetail(
                label: "\(side) marked OI value",
                value: leg.markedOiValue.map(notionalText) ?? "Unavailable"
            ),
        ]
    }

    private static func markedOiScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        (strike.call?.markedOiValue ?? 0) + (strike.put?.markedOiValue ?? 0)
    }

    private static func quoteSizeScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        let callSize = (strike.call?.bidSize ?? 0) + (strike.call?.askSize ?? 0)
        let putSize = (strike.put?.bidSize ?? 0) + (strike.put?.askSize ?? 0)
        return Double(callSize + putSize)
    }

    private static func openInterestScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        Double((strike.call?.openInterest ?? 0) + (strike.put?.openInterest ?? 0))
    }

    private static func volumeScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        Double((strike.call?.volume ?? 0) + (strike.put?.volume ?? 0))
    }

    private static func relativeSpreadScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        max(strike.call?.relativeSpread ?? 0, strike.put?.relativeSpread ?? 0)
    }

    private static func roundTripScore(_ strike: OptionsAnalyticsStrikeDTO) -> Double {
        max(strike.call?.roundTripCost ?? 0, strike.put?.roundTripCost ?? 0)
    }

    static func railWidth(for paneWidth: Double) -> Double {
        min(112, max(56, paneWidth / 100 * 28))
    }

    static func accessibilitySummary(
        snapshot: OptionsAnalyticsSnapshotDTO,
        settings: OptionsAnalyticsSettings,
        now: Date = Date()
    ) -> String {
        let quality = snapshot.quality
        var parts = [
            "Options Structure \(snapshot.scope.symbol) expiration \(snapshot.scope.expiration)",
            "Root \(snapshot.scope.rootSymbol), \(snapshot.scope.settlementStyle.rawValue.uppercased()) settlement",
            "Spot \(priceText(snapshot.scope.spot)), forward \(priceText(snapshot.scope.forward))",
            "Call gamma \(optionalNotionalText(snapshot.structure.callGammaExposure))",
            "Put gamma \(optionalNotionalText(snapshot.structure.putGammaExposure))",
            "Gross gamma \(optionalNotionalText(snapshot.structure.grossGammaExposure))",
            "Call delta notional \(optionalNotionalText(snapshot.structure.callDeltaNotional))",
            "Put delta notional \(optionalNotionalText(snapshot.structure.putDeltaNotional))",
            snapshot.exposureUnit,
            "\(quality.status.rawValue) status",
            "Coverage \(quality.coverage.contractsIncluded) of \(quality.coverage.contractsTotal) contracts, "
                + "\(Int((quality.coverage.ratio * 100).rounded()))%",
            "\(quality.feedMode.rawValue) feed",
            "Cache \(quality.cacheStatus.rawValue)",
            "calculation \(quality.calculationVersion)",
            "OI effective \(quality.oiEffectiveDate ?? "unavailable")",
        ]
        appendAge(named: "observed", timestamp: snapshot.scope.observedAt, now: now, to: &parts)
        appendAge(named: "quote", timestamp: quality.quoteAsOf, now: now, to: &parts)
        appendAge(named: "Greeks", timestamp: quality.greeksAsOf, now: now, to: &parts)
        appendSettlement(snapshot.scope.settlementAt, now: now, to: &parts)
        if let range = snapshot.impliedRange, settings.showImpliedRange {
            parts.append(
                "\(range.label) \(priceText(range.lower)) to \(priceText(range.upper))"
            )
            parts.append(
                "straddle breakevens \(priceText(range.straddleLower)) to \(priceText(range.straddleUpper))"
            )
        }
        if let wall = snapshot.structure.callWall {
            parts.append("Call wall \(priceText(wall))")
        }
        if let wall = snapshot.structure.putWall {
            parts.append("Put wall \(priceText(wall))")
        }
        if let strike = snapshot.structure.maxOpenInterestStrike {
            parts.append("Max OI strike \(priceText(strike))")
        }
        if let concentration = snapshot.structure.grossGammaConcentration {
            parts.append("Gross gamma concentration \(Int((concentration * 100).rounded()))%")
        }
        let legs = snapshot.strikes.flatMap { [$0.call, $0.put].compactMap({ $0 }) }
        if settings.showGammaProfile {
            let callLegs = snapshot.strikes.lazy.filter { $0.call != nil }.count
            let putLegs = snapshot.strikes.lazy.filter { $0.put != nil }.count
            parts.append("Gamma profile, \(callLegs) call legs and \(putLegs) put legs")
        }
        if settings.showMarkedOi {
            let markedCount = legs.lazy.filter { $0.markedOiValue != nil }.count
            parts.append("Marked open interest layer, \(markedCount) contracts")
        }
        if settings.showLiquidity {
            let liquidityCount = legs.lazy.filter {
                $0.relativeSpread != nil || $0.roundTripCost != nil
            }.count
            parts.append("Liquidity layer, \(liquidityCount) contracts")
        }
        if settings.showDealerProxy, let proxy = snapshot.scenarios.callPutDealerProxy {
            parts.append("Assumption: \(proxy.assumption)")
            let roots = proxy.gammaRoots.isEmpty
                ? "unavailable"
                : proxy.gammaRoots.map(priceText).joined(separator: ", ")
            parts.append("Proxy roots \(roots)")
        }
        parts.append(contentsOf: quality.warnings.map { "Warning: \($0)" })
        return parts.joined(separator: ". ")
    }

    private static func appendAge(
        named name: String,
        timestamp: String?,
        now: Date,
        to parts: inout [String]
    ) {
        guard let timestamp,
              let date = DateParsing.dateTime(timestamp)
        else { return }
        let seconds = max(0, Int(now.timeIntervalSince(date).rounded()))
        parts.append("\(name) source \(timestamp), current age \(seconds) seconds")
        if name == "quote" {
            parts.append("quote age \(seconds) seconds")
        } else if name == "Greeks" {
            parts.append("Greek age \(seconds) seconds")
        }
    }

    private static func appendSettlement(
        _ timestamp: String,
        now: Date,
        to parts: inout [String]
    ) {
        guard let settlement = DateParsing.dateTime(timestamp) else { return }
        let interval = Int(settlement.timeIntervalSince(now).rounded())
        if interval >= 0 {
            parts.append("Settlement \(timestamp), in \(interval) seconds")
        } else {
            parts.append("Settlement \(timestamp), passed \(-interval) seconds ago")
        }
    }

    private static func priceText(_ value: Double) -> String {
        String(format: "%.2f", value)
    }
}
